heron/proto/ckptmgr.proto

```diff
@@ -0,0 +1,182 @@
+package heron.proto.ckptmgr;
+option java_package = "com.twitter.heron.proto.ckptmgr";
+option java_outer_classname = "CheckpointManager";
+
+import "common.proto";
+import "physical_plan.proto";
+
+// message stored in zk by the tmaster
+message StatefulConsistentCheckpoints {
+  // This is the most recent checkpoint that we believe exists
+  required string most_recent_checkpoint_id = 1;
+  // It might be possible that for various reasons(fs corruption, etc)
+  // the last checkpoint is not accessible. This is the set of
+  // other globally consistent checkpoints that can be recovered from.
+  // The order of preference(and recency) is from beginning to end
+  repeated string backup_checkpoint_ids = 2;
+}
+
+// tmaster -> stmgr messages
+
+// Message sent to stmgrs by the tmaster to initiate checkpointing
+message StartStatefulCheckpoint {
+  required string checkpoint_id = 1;
+}
+
+// Message sent by tmaster to stmgr asking them to reset their instances
+// to this checkpoint
+message RestoreTopologyStateRequest {
+  required string checkpoint_id = 1;
+  required int64 restore_txid = 2; 
+}
+
+// Message that stmgr sends to tmaster after it restores
+// all of its local instances to a checkpoint_id
+message RestoreTopologyStateResponse {
+  required heron.proto.system.Status status = 1;
+  required string checkpoint_id = 2;
+  required int64 restore_txid = 3; 
+}
+
+// Message sent by stmgr to tmaster asking it to reset the topology
+// to some valid checkpoint. This is sent either if stmgr dies
+// and comes back up or if an instance dies.
+message ResetTopologyState {
+  optional string dead_stmgr = 1;
+  optional int32 dead_taskid = 2;
+}
+
+// Message sent by stmgr to tmaster informing it about
+// the fact that we stored a checkpoint belonging
+// to the instance
+message InstanceStateStored {
+  required string checkpoint_id = 1;
+  required heron.proto.system.Instance instance = 2;
+}
+
+// Message sent by tmaster to stmgr to start processing
+message StartStmgrStatefulProcessing {
+  required string checkpoint_id = 1;
+}
+
+// tmaster -> ckptmgr messages
+
+// This is the message that a tmaster sends
+// when it wants to establish the connection
+// with checkpoint manager
+message RegisterTMasterRequest{
+  required string topology_name = 1;
+  required string topology_id = 2;
+}
+
+// This is the message that checkpoint manager
+// sends when it receives the register request
+// from tmaster
+message RegisterTMasterResponse {
+  required heron.proto.system.Status status = 1;
+}
+
+// Message sent by tmaster to ckptmgr to cleanup
+// old checkpoint state.
+message CleanStatefulCheckpointRequest {
+  // Any checkpoints older than this can be cleaned
+  optional string oldest_checkpoint_preserved = 1;
```

Comment:
This is just the oldest checkpoint that needs to be preserved. All older ones can be cleaned